CorsixTH/Src/midi_player.h

```diff
@@ -0,0 +1,309 @@
+      : midi_out(midi_out), port(port) {
+    if (port == midi_virtual_port_index) {
+      midi_out.openVirtualPort(midi_external_port_name);
+    } else {
+      midi_out.openPort(port, midi_external_port_name);
+    }
+  }
+  open_midi_port(const open_midi_port&) = delete;
+  open_midi_port& operator=(const open_midi_port&) = delete;
+  open_midi_port(const open_midi_port&&) = delete;
+  open_midi_port& operator=(const open_midi_port&&) = delete;
+
+  /**
+   * Release the open port
+   */
+  ~open_midi_port() { midi_out.closePort(); }
+
+  /**
+   * True if the port is virtual (midi_virtual_port_index)
+   */
+  bool is_virtual() const { return port == midi_virtual_port_index; }
+
+  /**
+   * The owned port index
+   */
+  unsigned int get_port() const { return port; }
+
+ private:
+  /// Reference to MIDI API
+  RtMidiOut& midi_out;
+
+  /// Owned/opened port index
+  unsigned int port;
+};
+
+/**
+ * Commands between the midi_player and the playback thread
+ */
+enum class player_command {
+  /// Stop MIDI playback
+  stop,
+
+  /// Pause MIDI playback with the intention to resume from the same point
+  pause,
+
+  /// Resume MIDI playback after pause
+  resume,
+
+  /// Adjust the master volume level, based on the currently set value in
+  /// midi_player
+  set_volume,
+
+  /// Continue processing as if there was no command
+  noop
+};
+
+/**
+ * A thread safe queue of player_command with optional blocking on pop
+ *
+ * For sharing commands between the playback thread and the main thread.
+ */
+class player_command_queue {
+ public:
+  /**
+   * Push a command to the back of the queue
+   * @param command The command to push
+   */
+  void push(player_command command);
+
+  /**
+   * Pop a command off the queue
+   *
+   * Waits if the queue is empty and blocking is set to true.
+   * @param blocking Whether to block and wait for a command if the queue is
+   *                 empty
+   * @return The command at the front of the queue, or noop if the queue is
+   *         empty.
+   */
+  player_command pop(bool blocking);
+
+  /**
+   * Empty the queue
+   */
+  void clear();
+
+ private:
+  /// Mutex to guard the queue
+  std::mutex mut{};
+
+  /// The queue
+  std::queue<player_command> queue{};
+
+  /// Condition to wait on when empty and notify when populated
+  std::condition_variable cv{};
+};
+
+/**
+ * Control class for playing MIDI
+ */
+class midi_player {
+ public:
+  /**
+   * Initialize a connection on a given MIDI API and port for playing CorsixTH
+   * XMI files
+   * @param api The MIDI API to use, one of:
+   *            - <empty string> or "Native" - Any available API
+   *            - "ALSA" - Linux ALSA
+   *            - "JACK" - JACK on Linux and Unix where available
+   *            - "CoreMIDI" - MacOS CoreMIDI
+   *            - "Windows MM" - Windows MIDI Api
+   *            - "Web MIDI API" - Web MIDI API
+   *            - "Windows UWP" - Universal Windows Platform MIDI API (RtMidi
+   * 6+)
```

Comment:
Line is just too long. Perhaps break the line after "API"?
(For some reason I cannot suggest that.